pyartcd/pyartcd/umb_client.py

```diff
@@ -0,0 +1,375 @@
+    async for message in receiver.iter_messages():
+        print(message.headers["message-id"])
+    ```
+    """
+
+    def __init__(self, id: str) -> None:
+        self.id = id
+        self._queue: asyncio.Queue[Optional[stomp.utils.Frame]] = asyncio.Queue()
+        self._loop = asyncio.get_event_loop()
+        self.closed = False
+
+    async def iter_messages(self):
+        while True:
+            r = await self._queue.get()
+            if r is None:  # None represents "EOF"
+                return
+            yield r
+
+    def put_frame(self, frame: stomp.utils.Frame):
+        if self.closed:
+            raise ValueError("AsyncUMBReceiver is closed")
+        self._queue.put_nowait(frame)
+
+    def close(self):
+        self.closed = True
+        self._queue.put_nowait(None)  # None represents "EOF"
+
+
+class UMBClientConnectionListener(stomp.ConnectionListener):
+    """ This class is used internally by AsyncUMBClient to handle stomp.py events.
+    """
+
+    def __init__(self, client: "AsyncUMBClient", print_to_log=True):
+        self._client = client
+        self._futures: Dict[str, asyncio.Future] = {}
+        self._receivers: Dict[str, AsyncUMBReceiver] = {}
+        self.print_to_log = print_to_log
+
+    def add_future(self, id: str, fut: asyncio.Future):
+        self.remove_future(id)
+        self._futures[id] = fut
+
+    def remove_future(self, id: str):
+        old = self._futures.pop(id, None)
+        if old and not old.done():
+            old.cancel()
+
+    def add_receiver(self, id: str, receiver: AsyncUMBReceiver):
+        self._receivers[id] = receiver
+
+    def remove_receiver(self, id: str, close=True):
+        old = self._receivers.pop(id, None)
+        if old and close:
+            old.close()
+
+    def __print(self, msg, *args):
+        if self.print_to_log:
+            logging.debug(msg, *args)
+        else:
+            print(msg % args)
+
+    def on_connecting(self, host_and_port):
+        """
+        :param (str,int) host_and_port:
+        """
+        self.__print("on_connecting %s %s", *host_and_port)
+
+    def on_connected(self, frame):
+        """
+        :param Frame frame: the stomp frame
+        """
+        self.__print("on_connected %s %s", frame.headers, frame.body)
+        self._complete_future("on_connected", None)
+
+    def _complete_future(self, id: str, result: Any):
+        fut = self._futures.get(id)
+        if not fut:
+            return
+        fut.get_loop().call_soon_threadsafe(fut.set_result, result)
+
+    def _err_future(self, id: str, err: Exception):
+        fut = self._futures.get(id)
+        if not fut:
+            return
+        fut.get_loop().call_soon_threadsafe(fut.set_exception, err)
+
+    def on_disconnected(self):
+        self.__print("on_disconnected")
+        for _, r in self._receivers.items():
+            # close the event loop of the sender thread
+            r._loop.call_soon_threadsafe(r.close)
```

Comment:
It would look better if we didn't have to access a protected member (and an important one!). How about setting a getter `@property`? Besides, don't we always have only one `asyncio` loop running? Do we really need to store a reference to it within  `AsyncUMBReceiver`? Wondering if we could just use `asyncio.get_event_loop()` where we need it

Reply:
I think I need to add more comments here to explain it.

stomp.py calls those Listener methods in a separated thread (started and managed by itself). That thread doesn't have a event loop.

There are 2 event loops here:
- the main event loop
- the event loop in the sender thread

I meant to close the event loop in the sender thread, so the code here is incorrect. I will fix it later.

Reply:
I added a new method `close_soon_threadsafe` to avoid using protected memeber.